### ISSUE TITLE
Skip fully black or white tiles when creating Deep Zoom pyramids?

### ISSUE THREAD

**Issue by perog** [jcupitt/libvips#352]
I have lots of large mirax (mrxs) slides (120000x260000pixels on level 0) with a lot of white areas between the samples on the slide. Is it possible to skip these tiles when creating the stack and by that lower the time needed to create these stacks as well as the total stack size that becomes shrunk by 80% (2.6GB to 0.5GB)? I use the stacks with openseadragon and missing tiles are handled there instead.

The vips command I run is 'vips dzsave example.mrxs[autocrop] example' and it takes over 30 minutes to create the stack. Is it expected to take such long time for these sizes? The server I run on is a quite new multicore linux server, not entirely sure about the specs though... (40 *
Intel(R) Xeon(R) CPU E5-2650 v3 @ 2.30GHz according to 'less /proc/cpuinfo') Could you advice me if there is some speed optimiziation that I can do?

**Comment by jcupitt** [comment]
It could skip blank tiles, but I don't think it would speed it up much. I believe google maps layout allows blank tiles, but I'm not sure if zoomify or deepzoom permit this. You're right, it would at least save some disk space. 

120k x 260k is large, I'd expect it to take a while. With 256x256 tiles, that's over 600,000 separate jpeg files, ouch. Even if it takes 30 minutes, that still means it's generating 350 jpeg files every second. 

You could turn the tile size up, that might speed it up a bit. An SSD would help too, probably. You could also experiment with zip output, it can be faster if your disk is very slow. 

Another option is to convert on the fly. So leave your mrxs files on disc, and serve tiles from them dynamically as clients request them. For small numbers of users, it's not too bad. I made a small on-the-fly server:

https://github.com/jcupitt/tilesrv

**Comment by jcupitt** [comment]
I had a quick hack:

https://github.com/jcupitt/libvips/tree/no-blank-tiles

It seems to work. In google maps mode, it now skips tiles which are equal to the background colour. I had some code in there earlier to do this but removed it as it was a lot slower. This version seems to be fast enough that you can probably leave it on all the time. Before:

```
$ time vips dzsave CMU-1.svs x --layout google --background "243 243 243 255"
real	1m1.940s
user	2m15.004s
sys	0m37.092s
```

After:

```
$ time vips dzsave CMU-1.svs x --layout google --background "243 243 243 255"
strip_work: skipping blank tile 176 x 67
strip_work: skipping blank tile 21 x 112
real	1m3.503s
user	2m16.012s
sys	0m40.328s
```

Though it only saves two tiles in this case.

Would you be able to test?

**Comment by jcupitt** [comment]
Hi again, did you get a chance to test this?

**Comment by jcupitt** [comment]
Closing unmerged. Please reopen if this is still an issue.

**Comment by bverem** [comment]
I would like to test this! However, I'm pretty new to any sort of programming. Can you point me to a resource that would explain how I would go about using the linked files to test this?

**Comment by jcupitt** [comment]
Hi @bverem, what platform are you using? It should be fairly easy to build on OS X and Linux, but rather challenging on Windows. I'll update that branch to latest.

**Comment by bverem** [comment]
Windows, sadly.

**Comment by jcupitt** [comment]
I'll make you a test binary, hang on.

**Comment by jcupitt** [comment]
Here's a test build of that branch:

http://www.vips.ecs.soton.ac.uk/development/8.3/win32/vips-dev-w64-no-blank-tiles-8.3.zip

It's got some extra diagnostics in. If I run this on an image with large white areas I see:

```
$ vips.exe dzsave big.jpg x --layout google
strip_work: skipping blank tile 0 x 0
strip_work: skipping blank tile 1 x 0
strip_work: skipping blank tile 4 x 0
strip_work: skipping blank tile 3 x 0
strip_work: skipping blank tile 5 x 0
strip_work: skipping blank tile 6 x 0
... etc.
```

So hopefully you can work out which tiles were not saved.

The questions I have would be:

1. How well does this work on real slides? It only doesn't save tiles which are exactly equal to the background colour, so even a tiny hint of noise means that you will get a real tile. Maybe there should be a threshold, so "don't save if the maximum difference from the background in a tile is less than X". But what's a good value for X? We'd need to run some tests. 

2. And of course it would be good to verify that the pyramids it saves actually work with real images on a real server.

**Comment by bverem** [comment]
So here's my workflow at the moment: I get Mirax files scanned by a 3DHistech machine, but in the metadata for those files, it specifies only three tiers, so vips also only processes three tiers and the deeper levels don't get processed. As a workaround, I've been using 3DHistech's file management program to export these files as JPEG-compressed TIFF files, which can then be used in vips to convert to Google Maps files.

Using this method, I cropped as much as I could from one of my slides, converted to a JPEG-compressed TIFF with the "quality" set at 100/100, and used vanilla vips as well as the test build. The original file size was 1.53 GB for the TIFF, and with vanilla vips, the GMaps-ready file size was 412MB (545MB on disk) containing 62,342 files. This build reduced the file size to 377MB (458 on disk) and the number of files to 40,223. Looking through the command prompt window, there are _many_ skipped blank tiles.

I uploaded these to my bucket in Google Cloud, and have verified that they kind of work. See [here](http://brandonveremis.com/17_100.html) for the slide without removal of blank tiles, and [here](http://brandonveremis.com/17_blank.html) for the slide with blank tiles removed. It does not appear that blank.png is being loaded in the missing tile areas. I would be much obliged if you had any ideas as to why not (my programming skills are minimally existent).

I was also able to use the .zip output without any issues, reducing the file size to 383MB. 

The benefit of skipping these tiles is potentially huge. While skipping a file doesn't really save much time in converting these files, it saves a fair amount of disk storage, and since I have a lesser number of files to upload, Google charges me less for the upload actions (though this has fewer implications for those who have institutions with large amounts of storage space). It is, however, important to crop as much out as possible, as well as use good images. This will probably work better if the slides are as clean as possible before scanning, and if the machine doesn't properly white-balance the background (which happens sometimes with our scanning software), they should be re-scanned. I don't think this technique will work well on a gray background.

As for future iterations, it would be useful to try setting certain thresholds. As I manually look through which tiles were retained, many are visually entirely white, likely the white tiles that did load in the second link. With slide images, it might be hard to draw a line between a mostly white background tile and a mostly white tissue tile (such as in a fatty tumor). I do have some slides that contain such white-heavy tissues, and I could use those for testing purposes, or I could just continue to use the specimen above and look at the white-heavy areas.

At this point, I think it would be useful to:

1. Establish why blank.png is not being loaded (I assume it's an oversight on my part).
2. Increase the sensitivity for what counts as a "blank" tile, and test on white-heavy images. Alternatively, we could just use the specimen linked above, and look at the white-heavy areas in the tissue (fat and salivary glands).
3. After it works well on routine H&E slides like the linked specimen, we could test on other, possibly trickier types of specimens like immunohistochemistry (which are often very light-staining).

**Comment by bgilbert** [comment]
> So here's my workflow at the moment: I get Mirax files scanned by a 3DHistech machine, but in the metadata for those files, it specifies only three tiers, so vips also only processes three tiers and the deeper levels don't get processed.

It's possible that VIPS isn't actually reading the slide data, but only the JPEG thumbnail inside the `.mrxs` file itself.  Could you run `vipsheader` on the `.mrxs` file and report whether it says `jpegload` or `openslideload`?

**Comment by jcupitt** [comment]
Hi @bgilbert, yes, you're correct, this was a bug in vips 8.2.1, mirax images were being opened as jpegs. 

@bverem, that test build should be able to open mirax images directly without having to go via tiff.

Your test image upload is just coming up as empty for me:

http://brandonveremis.com/17_blank.html

I see a plain mid-grey screen, no detail at all. I tried in firefox and chrome. 

It's great it seems to be working, thank you for testing. I don't know why blank.png is not being loaded, perhaps the google maps viewer needs to be told to use it for missing tiles? I don't see a reference to it in your init function. 

I'll see how easy it would be to add a threshold for blankness.

**Comment by bverem** [comment]
Just to clarify, @jcupitt, you cannot see any images at all? If so, I'm
wondering if there's some sort of regional restriction on my website or
with the data in Google Cloud. I'll look into that. Usually when the GMaps
UI loads without any sort of image information, it's because Google Maps is
unable to find the tiles. Typically, I see this when I accidentally
misspell the URL, but it's working on my other devices as well as at my
work desktop. I'll keep looking into what I have to enter into the function
to get the blank tile to load, but in the brief time I looked yesterday, I
didn't find anything promising in the GMaps reference pages.

I will try to use the test build with mirax files directly (thank you also,
@bgilbert). I have a lot to do this week at my "day job", so I don't know
how much I can do immediately, but at the very least I'll be looking into
trying the test build along with trying to get blank tiles to load...
probably this upcoming weekend at the latest.

**Comment by jcupitt** [comment]
@bverem, I see a plain mid-grey page, with the google logo at the bottom-left, and streetview/zoom/terms at the bottom right, but no image data.

I've done a new version of the no-blank-tiles build:

http://www.vips.ecs.soton.ac.uk/development/8.3/win32/vips-dev-w64-no-blank-tiles-8.3a.zip

The diagnostic messages have gone, and it now uses a threshold to test for blankness. A tile is skipped if the maximum absolute difference between it and blank.png is 5 or less. I wondered about the value to use, 10 might work better, but could possibly produce very slight lines at tile edges. 5 should be invisible to anyone.

**Comment by bverem** [comment]
Hmmm, I get the same thing when I access the site from my workplace
desktops, but on non-institutional devices, it appears to work.
Interestingly, many other digital slide sites (including those that don't
use GMaps) don't work on these institutional devices, but I'm not sure why.
I'll try to find an answer for you and get it fixed.

I'll test that new build sometime towards the end of the week, at least by
the weekend. Thank you again for your hard work on this! I think it'll
really help reduce file size substantially.

**Comment by bverem** [comment]
I attempted using the test build on the same file; it only reduced the number of blank tiles from 40,223 to 40,211. Additionally, please let me know if the aforementioned links are working for you now. I cannot yet figure out how to have Google fill in the empty tiles with blank.png, so I'm still working on that.

**Comment by jcupitt** [comment]
Hmm, strange, you'd think it would make more difference. At least the number of tiles dropped. 

Sorry @bverem, your maps demo is still not working for me, I have broken tiles everywhere.

**Comment by felixbuenemann** [comment]
This is a great feature I'll have to try it out. I have product images on a white backdrop and JPEG's Huffman coding is not good at compressing those "empty" tiles. It also saves lot's of unneeded HTTP requests, although that's not a big issue over HTTP/2.

```sh
# create a completely white jpeg at quality 75
convert -size 1024x1024 xc:white -type TrueColor canvas.jpg
# try to gzip it, to see how efficient compression was
gzip -k9 canvas.jpg
# compare
ls -hal canvas.jpg*
-rw-r--r--  1 felix  staff    12K  9 Mai 14:45 canvas.jpg
-rw-r--r--  1 felix  staff   203B  9 Mai 14:45 canvas.jpg.gz
```

*Anyone knows or tried if openseadragon supports the google layout out of the box?*

(I've been using the deepzoom layout so far.)

**Comment by jcupitt** [comment]
You should be able to make openseadragon use google layout if you define your own `getTileUrl`. 

I don't know if you can make it display `blank.png` for missing tiles. It looks like it supports sparse pyramids, but I can't see a simple example of how to set it up.

**Comment by felixbuenemann** [comment]
There's a method `tileExists` that could do the trick, if the metadata file allows to detect which tiles are missing.

**Comment by jcupitt** [comment]
No, sadly not. Google maps displays `blank.png` for all tiles, then does a fetch to get the image to display. If the fetch fails (because that tile is missing), you just see the blank. So the metadata does not record which tiles have been skipped.

**Comment by bverem** [comment]
Alright! I'm back now with some updates.

@jcupitt, I figured out why you weren't able to see that file: I evidently have to enable which files I want to be public in Google Cloud Storage. I haven't changed those files to public (it takes a long time to change the status of 40,000 files, even from the gsutil command line), but if you'd like to see the final product, let me know and I'd be happy to do that for you.

Second, you don't need to use the blank.png for the blank tiles. I simply set the Maps background to white. This also saves on server operations (not to mention a lot of hair-pulling on my end). If someone wanted to use a different custom tile that wasn't a solid color, it starts to get tricky. The suggestion I was given in stackoverflow was to use a custom 404 image, but as I don't have that ability in Google Cloud Storage, I can't test that. I don't think there's much demand for a non-solid color image, anyway.

So, I think this setup is a go! It seems to work well, and I don't think we need to work more on eliminating more tiles. Looking through what tiles were left behind, they seem to be on the periphery of the colored tiles, so they have a somewhat off-white appearance, but I don't think it's worth it to try to weed those out, too.

**Comment by felixbuenemann** [comment]
I think I saved roughly 50% in my tests. I have the files on another computer at work, so I can't check right now.

**Comment by bverem** [comment]
So another mild request, now that I have these files working. Would it be possible to make the background off-white images as "blank" in [this file](http://brandonveremis.com/slides/case3.html)? Since it's a little further than pure white, would there be a way to set the sensitivity of the blank tile removal component of vips in, say, the parameters? Sometimes, our scanner is really good at making the background white... other times, not so much. 

As before, let me know if this file doesn't work. It should now, though.

**Comment by jcupitt** [comment]
You can do this now: just set whatever colour you want as "background". For example:

```
vips dzsave huge.tif x --background 249
```

You can set a colour background too, perhaps:

```
vips dzsave huge.tif x --background "249 249 240"
```

**Comment by jcupitt** [comment]
I think this is all done, I'll close.

**Comment by bverem** [comment]
I'm back with another question! I've progressed from converting single whole slide images to a full stack application containing a thousand or so images, and I started using the DeepZoom format rather than Google Earth. Is there a similar method to removing the "white" tiles in DeepZoom images? I've upped the tile size to help save on storage, but skipping white tiles would be helpful in DZ as well.

Thank you again for this library; it's been supremely helpful in assembling this application!
